src/util/permissions.tsx

```diff
@@ -0,0 +1,377 @@
+  permissions?: LxdPermission[];
+  imageNamesLookup: Record<string, string>;
+  identityNamesLookup: Record<string, string>;
+}) => {
+  const { resourceType, permissions, identityNamesLookup, imageNamesLookup } =
+    args;
+  if (!permissions || !resourceType) {
+    return [];
+  }
+
+  const resourceOptions: OptionHTMLAttributes<HTMLOptionElement>[] = [
+    defaultOption,
+  ];
+
+  const processedResources = new Set<string>();
+  for (const permission of permissions) {
+    const resourceName = extractResourceNameFromUrl({
+      resourceType,
+      url: permission.url,
+      imageNamesLookup,
+      identityNamesLookup,
+    });
+
+    if (processedResources.has(resourceName)) {
+      continue;
+    }
+
+    processedResources.add(resourceName);
+    resourceOptions.push({
+      value: permission.url,
+      label: extractResourceNameFromUrl({
+        resourceType,
+        url: permission.url,
+        imageNamesLookup,
+        identityNamesLookup,
+      }),
+    });
+  }
+
+  resourceOptions.sort(sortOptions);
+
+  return resourceOptions;
+};
+
+export const generateResourceEntitlementOptions = (
+  resourceType: string,
+  permissions?: LxdPermission[],
+) => {
+  if (!permissions || !resourceType) {
+    return {};
+  }
+
+  const resourceEntitlements: Record<
+    string,
+    OptionHTMLAttributes<HTMLOptionElement>[]
+  > = {};
+
+  for (const permission of permissions) {
+    const resource = permission.url;
+    if (!resourceEntitlements[resource]) {
+      resourceEntitlements[resource] = [defaultOption];
+    }
+
+    resourceEntitlements[resource].push({
+      value: permission.entitlement,
+      label: permission.entitlement,
+    });
+  }
+
+  for (const resource in resourceEntitlements) {
+    // split entitlement options into two based on 'can_' prefix
+    const genericEntitlements: OptionHTMLAttributes<HTMLOptionElement>[] = [];
+    const granularEntitlements: OptionHTMLAttributes<HTMLOptionElement>[] = [];
+
+    for (const entitlementOption of resourceEntitlements[resource]) {
+      const entitlementValue = entitlementOption.value as string;
+      if (entitlementValue.includes("can_")) {
+        granularEntitlements.push(entitlementOption);
+        continue;
+      }
+
+      genericEntitlements.push(entitlementOption);
+    }
+
+    genericEntitlements.sort(sortOptions);
+    granularEntitlements.sort(sortOptions);
+
+    // add disabled option as a delimiter between the two sets of entitlements
+    if (genericEntitlements.length > 1 && granularEntitlements.length) {
+      genericEntitlements.splice(1, 0, {
+        disabled: true,
+        label: "Built-in entitlements",
+        value: "",
+      });
+
+      granularEntitlements.unshift({
+        disabled: true,
+        label: "Granular entitlements",
+        value: "",
+      });
+    }
+    resourceEntitlements[resource] =
+      genericEntitlements.concat(granularEntitlements);
+  }
+
+  return resourceEntitlements;
+};
```

Comment:
I think the available entitlements only dependant on the resource type. The entitlements will be the same for whichever resource is chosen. I wonder why we return the entitlements for each available resource.

Wouldn't it suffice to look at just one resource and use its entitlements for all other resources of the same resource type? Then we can simplify this function and the interface by just returning a bare array of entitlements.

Reply:
That's a good point, I did think about that but thought I'd rather not make that assumption in case if there are some edge cases that breaks this assumption. Since I have now looked at the data for all resource types extensively, it seems this assumption does hold true, so will update accordingly.